core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -505,7 +525,29 @@ public void refresh(String source) throws EngineException {
             throw new RefreshFailedEngineException(shardId, t);
         }
 
-        // TODO: maybe we should just put a scheduled job in threadPool?
+        mergeScheduler.refreshConfig();
+
+        // we must also refresh our internal searcher, so that a subsequent real-time get (which uses internal searcher) comes from the
+        // index and not xlog:
+        refreshInternal();
+    }
+
+    private void refreshInternal() throws EngineException {
+        // we obtain a read lock here, since we don't want a flush to happen while we are refreshing
+        // since it flushes the index as well (though, in terms of concurrency, we are allowed to do it)
+        try (ReleasableLock lock = readLock.acquire()) {
```

Comment:
do we want to unify this code with the refresh method by making this get a manager to work on? (+ a string description for failures)